### PR TITLE
feat(gbench): Adds gbench of GetSpeed with traffic data

### PR DESCRIPTION
**Pull request by PureW**
Also breaks out test-utility functions into `test/util/traffic_utils.h

### CODE REVIEW

File: bench/CMakeLists.txt [hunk 1]
```diff
@@ -32,7 +34,12 @@ macro (add_valhalla_benchmark target_file)
     COMMENT "Running ${target_name} in ${CMAKE_CURRENT_BINARY_DIR}"
     DEPENDS ${target_name})
   set_target_properties(run-${target_name} PROPERTIES FOLDER "Benchmarks")
+  add_dependencies(run-${target_name} utrecht_tiles)
```
Comment (PureW) [inline bench/CMakeLists.txt:37]: The new `routes` bench uses utrecht data.

File: bench/thor/costmatrix.cc [hunk 2]
```diff
@@ -106,7 +108,7 @@ static void BM_UtrechtCostMatrix(benchmark::State& state) {
     thor::CostMatrix matrix;
     auto result =
         matrix.SourceToTarget(sources, sources, reader, &costs, sif::TravelMode::kDrive, 100000.);
-    result_size = result.size();
+    result_size += result.size();
```
Comment (PureW) [inline bench/thor/costmatrix.cc:111]: I think this was a bug in the previous code

File: test/gurka/gurka.h [hunk 3]
```diff
@@ -515,27 +515,6 @@ inline void build_pbf(const nodelayout& node_locations,
...
 
-/**
- * Generate a new GraphReader that doesn't re-use a previously
- * statically initizalized tile_extract member variable.
- *
- * Useful if you need to reload a tile extract within the same
- * process
- */
-std::shared_ptr<valhalla::baldr::GraphReader>
-make_clean_graphreader(const boost::property_tree::ptree& pt) {
```
Comment (PureW) [inline test/gurka/gurka.h:526]: I'm moving this into the new `test/util/traffic_utils.h` file as it's useful outside of gurka tests.

File: test/gurka/test_traffic.cc [hunk 4]
```diff
@@ -14,129 +15,6 @@
...
-  char checksum[8];
-  char type;
-  char linkname[100];
-  char _padding[255];
-} mtar_raw_header_t_;
-
-/*************************************************************/
-
-// Helper function for updating the BD edge in the following test
-void update_all_edges_but_bd(const valhalla::gurka::map& map,
```
Comment (PureW) [inline test/gurka/test_traffic.cc:34]: Moving a more general implementation of this into the new `test/util/traffic_utils.h`

File: test/util/traffic_utils.h [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    mtar_t tar;
+    auto tar_open_result = mtar_open(&tar, traffic_extract.c_str(), "w");
+    if (tar_open_result != MTAR_ESUCCESS) {
+      throw std::runtime_error("Could not create traffic tar file");
+    }
+
+    baldr::GraphReader reader(config.get_child("mjolnir"));
+
+    auto tile_ids = reader.GetTileSet();
+    // Traffic data works like this:
```
Comment (mookerji) [inline test/util/traffic_utils.h:76]: Should this explanation and the utility around reading/writing traffic tar files/tiles be in the main libvalhalla?

File: bench/thor/routes.cc [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+namespace {
+
+void create_costing_options(Options& options) {
+  const rapidjson::Document doc;
+  sif::ParseAutoCostOptions(doc, "/costing_options/auto", options.add_costing_options());
+  options.add_costing_options();
+}
+
+boost::property_tree::ptree json_to_pt(const std::string& json) {
```
[the remark this answers is not in the document]
Comment (mookerji) [reply bench/thor/routes.cc:26]: See https://github.com/valhalla/valhalla/pull/2485/files#r460342470

File: valhalla/baldr/traffictile.h [hunk 7]
```diff
@@ -96,6 +100,30 @@ struct TrafficSpeed {
         throw std::logic_error("Bad subsegment");
     }
   }
+
```
[the remark this answers is not in the document]
Comment (PureW) [reply valhalla/baldr/traffictile.h:103]: Ahh these are the subsegments. I realise we didn't add doc-comments describing this new feature to the header. Each edge can be split in up to three sections that each can describe different live traffic. I'll make a todo to add docs around this.

The "unit" of the fields is not a simple one, it is described in the header file. For the speed fields, each increment is 2kph (i.e. the value is left shifted once to get kph).

File: test/util/traffic_utils.h (hunk 5 above)
Comment (PureW) [reply test/util/traffic_utils.h:76]: Perhaps the explanation could also live in the graphreader but the code itself here is not meant for production and should stay on the testing side of the code.

File: bench/thor/routes.cc [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+  if (origins.size() == 0) {
+    throw std::runtime_error("No origins available for test");
+  }
+
+  std::size_t route_size = 0;
+
+  for (auto _ : state) {
+    thor::BidirectionalAStar astar;
+    for (int i = 0; i < origins.size(); ++i) {
+      // LOG_WARN("Running index "+std::to_string(i));
```
Comment (mookerji) [inline bench/thor/routes.cc:176]: nit: stray comment.

File: valhalla/baldr/traffictile.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (mookerji) [reply valhalla/baldr/traffictile.h:104]: > I'll make a todo to add docs around this.

Great!

> The "unit" of the fields is not a simple one, it is described in the header file. For the speed fields, each increment is 2kph (i.e. the value is left shifted once to get kph).

Is the reason around this complication that we're relying on bitpacking into a field here for ... what purpose?

File: bench/thor/routes.cc [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+  };
+  valhalla_tests::utils::customize_live_traffic_data(config, generate_traffic);
+}
+
+BENCHMARK(BM_UtrechtBidirectionalAstar)->Unit(benchmark::kMillisecond);
+
+/** Benchmarks the GetSpeed function */
+static void BM_GetSpeed(benchmark::State& state) {
+
+  const auto config = build_config("get-speed.tar");
```
[the remark this answers is not in the document]
Comment (mookerji) [reply bench/thor/routes.cc:215]: Passing on this for now.

File: valhalla/baldr/traffictile.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (PureW) [reply valhalla/baldr/traffictile.h:104]: Space, which relates to performance as a bigger structure means we can squeeze fewer edges into the cpu caches.

**Pull request merged**